[WIP][INT-4566 ]-Implement Inbound channel adapter for spring data r2dbc

rohanmukesh opened the pull request:
I have added Inbound channel adapter for spring data r2db.
The inbound adapter (R2dbcMessageSource) allows users to send query in two forms:
 a) String -> This uses databaseclient.execute(..) to execute sql query and return result.
 b) Query Object -> This uses r2dbcEntityOperations.select(..) to execute Query and return result.

I have added Test Class: R2dbcMessageSourceTests with 8 test cases. The test cases tests all the scenarions which includes:
String query Test with Single result or All Results, Query Object Test with Single Result or All Results and check for bad sql grammer, IllegalStateException.

rohanmukesh committed:
[WIP][INT-4566]-R2DBC Inbound Channel Adapter- Implementation of Select

artembilan reviewed (changes requested):
Would you mind to rebase your branch to the latest `upstream master`?
Plus squash all these commits: we need separate commits only after review.
Right now I see some unrelated commits in this PR. Probably because your local `master` is not fresh according to upstream...

artembilan commented on spring-integration-r2dbc/src/main/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSource.java:
```diff
@@ -0,0 +1,212 @@
+	private Object executeQuery(Query query) {
+		Object result;
+		if (this.expectSingleResult) {
+			result = this.r2dbcEntityOperations.selectOne(query, this.entityClass);
```
I think we should not rely on the `@Table` on the provided class.
I really believe that in most cases people are going to deal with just a `HashMap`.
Just because we interact here with messaging system and it is really hard to carry a particular domain class from one endpoint to another. 

So, I think we should get rid of `r2dbcEntityOperations` contract over here and just deal with the plain `DatabaseClient`. This way end-user will be able to not tie his/her logic with persistence model and those extra annotations.

We probably need to revise the `Query`  logic then in favor of `SelectSpec`...

artembilan commented on spring-integration-r2dbc/src/test/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSourceTests.java:
```diff
@@ -0,0 +1,280 @@
+ *  @since 5.4
+ */
+@SpringJUnitConfig (R2dbcIntegrationTestConfiguration.class)
+public class R2dbcMessageSourceTests {
```
`@DirtiesContext`, please, to stop the application context and clean up the memory from that H2 DB.

artembilan reviewed (changes requested):
Some review so far.
Thanks for the effort, but I know this is not an easy task.
So, let's engineer the final solution together!

rohanmukesh committed:
Fixed Review comments

rohanmukesh committed:
Fixed review comments

rohanmukesh commented:
Hello Artem, I have updated pull request with changes you requested and also with only "String Query". Although I have kept the constructor for queryExpression.  Removed the dependency on Entity @Table and just using databaseclient execute method for now. Let me know your feedback.

rohanmukesh committed:
Merge branch 'master' into feature/r2dbc-inbound-adapter

artembilan commented on spring-integration-r2dbc/src/test/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSourceTests.java:
```diff
@@ -0,0 +1,232 @@
+		Message<Person> message = MessageBuilder.withPayload(this.createPerson("Bob", 35)).build();
+		waitFor(handler.handleMessage(message));
+
+		message = MessageBuilder.withPayload(this.createPerson("Tom", 40)).build();
```
Don't use `this.` for method calls

artembilan reviewed (commented):
Please, be careful here in GH comments when you try to use something with `@` symbol. That "table" is a user here on GH and now he/she gets all the notifiations from our communication over here 😄 .
See GH comments formatting recommendations: https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks

rohanmukesh committed:
Fixed review comments

rohanmukesh committed:
Fixed review comments

artembilan commented on spring-integration-r2dbc/src/main/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSource.java:
```diff
@@ -0,0 +1,212 @@
+	 * which should resolve to a Relational 'query' string.
+	 * It assumes that the {@link R2dbcEntityOperations} is fully initialized and ready to be used.
+	 * The 'queryExpression' will be evaluated on every call to the {@link #receive()} method.
+	 *
```
Not addressed.

artembilan commented on spring-integration-r2dbc/src/main/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSource.java:
```diff
@@ -0,0 +1,188 @@
+								"or org.springframework.data.relational.core.query.Query, but not: " + value);
+					}
+					return result;
+				}).block();
```
You know I have changed my mind a little bit and figure out something like this for better optimization:
```
Mono<?> queryMono = Mono.fromSupplier(() -> this.queryExpression.getValue(this.evaluationContext));
if (this.this.expectSingleResult) {
     return queryMono.flatMap(this::selectOne)
}
else {
     return queryMono.flatMapMany(this::select)
}
```

The `selectOne` should return a `Mono` and `select` a `Flux`

rohanmukesh committed:
Fixed review comments

artembilan commented on spring-integration-r2dbc/src/main/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSource.java:
```diff
@@ -0,0 +1,185 @@
+		return queryMono.flatMapMany(this::select);
+	}
+
+	private Mono<?> selectOne(Object queryMono) {
```
Well, its not `Mono` here any more. Isn't it? 
I mean the name for the param is misleading 😄

artembilan commented on spring-integration-r2dbc/src/test/java/org/springframework/integration/r2dbc/inbound/R2dbcMessageSourceTests.java:
```diff
@@ -0,0 +1,169 @@
+		@Autowired
+		DatabaseClient databaseClient;
+
+		@Bean(name = "r2dbcMessageSourceSelectOne")
```
FYI: the method name becomes a bean name by default.
So, probably for this use-case you definitely don't need a `name` attribute on this annotation...

artembilan commented on spring-integration-r2dbc/src/test/java/org/springframework/integration/r2dbc/outbound/R2dbcMessageHandlerTests.java:
```diff
@@ -29,79 +27,49 @@
- *  @since 5.4
+ * @since 5.4
  */
 @SpringJUnitConfig
```
The `@DirtiesContext` here as well to close an application context and shoultdown embedded DB altogether.

artembilan reviewed (changes requested):
Very close to the merge!

artembilan reviewed (changes requested):
Please, consider to rebase to the latest `master` so Travis for your PR won't fail.

rohanmukesh committed:
Fixed Review comments

artembilan reviewed (changes requested):
Good. Let me pull it locally for final review and if there are only minor concerns I'll fix them and merge

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/eb3c6a017f2df2d954de1a842a22b2a15071c060.

Pay attention, please, how I have optimized a fetch preparation code.
